### PR TITLE
Memory allocator tests and benchmarks.

### LINKED ISSUE #289: Improve the jemalloc-based allocator implementation.
**Describe your problem**

The `TCACHE` is disabled in the jemalloc-base allocator, which, brings performance penality to the allocator and requires more work to be fixed.

Besides, currently the usage jemalloc seems has problems on Linux (Ubuntu 20.04) to reclaim allocated virtual memory 
address spaces.

**Additional context**

N/A.

### PR DESCRIPTION
**Pull request by peilii**
What do these changes do?
-------------------------

Memory allocator tests and benchmarks.

Related issue number
--------------------

Fixes #289

### CODE REVIEW

**Comment by sighingnow** [comment]
Hi @peilii,

Thanks for the effort. Could you please rename the directory to `benchmark`? `bench-x` looks a bit strange.

**Comment by peilii** [comment]
> Hi @peilii,
> 
> Thanks for the effort. Could you please rename the directory to `benchmark`? `bench-x` looks a bit strange.

No problem at all!

**Review by sighingnow — CHANGES_REQUESTED** [review]

File: src/common/memory/jemalloc.cc [hunk 1]
```diff
@@ -72,16 +72,27 @@ void* Jemalloc::Init(void* space, const size_t size) {
...
+  hooks_key << "arena." << std::to_string(arena_index_) << ".extent_hooks";
+  size_t len = sizeof(extent_hooks_);
+  if (auto ret = vineyard_je_mallctl(hooks_key.str().c_str(), &extent_hooks_,
+                                     &len, nullptr, 0)) {
+    int err = std::exchange(errno, ret);
+    PLOG(ERROR) << "Failed to create arena";
+    errno = err;
+    return nullptr;
+  }
+
```
[the remark this answers is not in the document]
Comment (peilii) [reply src/common/memory/jemalloc.cc:95]: No problem. I will look into what's going on.

[the remark this answers is not in the document]
Comment (peilii) [reply src/common/memory/jemalloc.cc]: Hi @sighingnow,
I would revert the change made in jemalloc.cc for this pull request, and find out what's going on in the next couple of days.

**Comment by sighingnow** [comment]
Please fix the format & lint error.

**Comment by peilii** [comment]
> Please fix the format & lint error.

I suppose it's done now. Should I also fix the Codacy Static Code Analysis problem?

**Comment by sighingnow** [comment]
>  Should I also fix the Codacy Static Code Analysis problem?

It doesn't matter.

**Review by sighingnow — APPROVED** [review]
LGTM.

**Comment by sighingnow** [comment]
Merged. Thanks!

**Comment by peilii** [comment]
> Merged. Thanks!

No problem! Thanks!

**Pull request merged**